src/main/groovy/life/qbic/portal/sampletracking/trackinginformation/query/sample/QuerySample.groovy

```diff
@@ -11,32 +14,55 @@ import life.qbic.portal.sampletracking.trackinginformation.query.locations.Query
+  final OpenBisClient openbisClient
+  final SampleTrackingQueryDataSource sampleTrackingDataSource
+  final Set<String> openbisSpaces
 
-    private QuerySampleOutput querySampleOutput
+  private QuerySampleOutput querySampleOutput
 
-    private QuerySample() {
-        // Avoids accidental default constructor call from within the class
-        throw new AssertionError()
-    }
+  private QuerySample() {
+    // Avoids accidental default constructor call from within the class
+    throw new AssertionError()
+  }
+
+  QuerySample(SampleTrackingQueryDataSource dataSource, QuerySampleOutput querySampleOutput, OpenBisClient openbisClient, Set<String> spaces) {
+    this.sampleTrackingDataSource = dataSource
+    this.querySampleOutput = querySampleOutput
+    this.openbisClient = openbisClient;
+    this.openbisSpaces = spaces;
+  }
 
-    QuerySample(SampleTrackingQueryDataSource dataSource, QuerySampleOutput querySampleOutput) {
-        this.sampleTrackingDataSource = dataSource
-        this.querySampleOutput = querySampleOutput
+  private void userIsAuthorized(String sampleId) {
+    List<ch.systemsx.cisd.openbis.generic.shared.api.v1.dto.Sample> samples =
+        openbisClient.getParentsBySearchService(sampleId)
+    if (samples.isEmpty()) {
+      throw new OpenbisEmptyProjectException("User tried searching for $sampleId, but related project does not have samples in openBIS")
     }
+    if (!openbisSpaces.contains(samples.get(0).getSpaceCode())) {
+      throw new OpenbisAuthorizationException("User tried searching for $sampleId, but user is not allowed to view this project.")
+    }
+  }
+
+  @Override
+  def querySampleById(String sampleId) {
+    try {
+      userIsAuthorized(sampleId)
 
-    @Override
-    def querySampleById(String sampleId) {
-        try {
-            Location sampleLocation = sampleTrackingDataSource.currentSampleLocation(sampleId)
-            Sample sample = new Sample()
-            sample.setCurrentLocation(sampleLocation)
-            sample.setCode(sampleId)
-            this.querySampleOutput.publishSample(sample)
-        } catch (SampleTrackingQueryException e) {
-            log.error("Query ${sampleId} failed.", e)
-            // We pass the original error message with a generic notification to the output port
-            this.querySampleOutput.invokeOnError "Could not retrieve information for sample $sampleId. ${e.message}"
-        }
+      Location sampleLocation = sampleTrackingDataSource.currentSampleLocation(sampleId)
+      Sample sample = new Sample()
+      sample.setCurrentLocation(sampleLocation)
+      sample.setCode(sampleId)
+      this.querySampleOutput.publishSample(sample)
+    } catch (SampleTrackingQueryException e) {
+      log.error("Query ${sampleId} failed.", e)
+      // We pass the original error message with a generic notification to the output port
+      this.querySampleOutput.invokeOnError "Could not retrieve information for sample $sampleId. ${e.message}"
```

Comment:
Do you really want the user to see the Exception message?

Reply:
good point, but that was already in the code before :)